util/ksonnet/ksonnet.go

```diff
@@ -132,12 +139,63 @@ func (k *ksonnetApp) Show(environment string) ([]*unstructured.Unstructured, err
 		if err != nil {
 			return nil, fmt.Errorf("Failed to unmarshal manifest from `ks show`")
 		}
+		err = remarshal(&obj)
+		if err != nil {
+			return nil, err
+		}
 		objs = append(objs, &obj)
 	}
 	// TODO(jessesuen): we need to sort objects based on their dependency order of creation
 	return objs, nil
 }
 
+// remarshal checks resource kind and version and re-marshal using corresponding struct custom marshaller. This ensures that expected resource state is formatter same as actual
+// resource state in kubernetes and allows to find differences between actual and target states more accurate.
+func remarshal(obj *unstructured.Unstructured) error {
+	var newObj interface{}
+	switch obj.GetAPIVersion() + ":" + obj.GetKind() {
+	case "apps/v1beta1:Deployment":
+		newObj = &v1beta1.Deployment{}
+		break
+	case "apps/v1beta2:Deployment":
+		newObj = &v1beta2.Deployment{}
+		break
+	case "extensions/v1beta1":
+		newObj = &v1ExtBeta1.Deployment{}
+		break
+	case "apps/v1beta1:StatefulSet":
+		newObj = &v1beta1.StatefulSet{}
+		break
+	case "apps/v1beta2:StatefulSet":
+		newObj = &v1beta2.StatefulSet{}
+		break
+	case "v1:Service":
+		newObj = &corev1.Service{}
+		break
+	}
+	if newObj != nil {
+		oldObj := obj.Object
+		data, err := json.Marshal(obj)
+		if err != nil {
+			return err
+		}
+		err = json.Unmarshal(data, newObj)
+		if err != nil {
+			return err
+		}
+		data, err = json.Marshal(newObj)
+		if err != nil {
+			return err
+		}
+		err = json.Unmarshal(data, obj)
+		if err != nil {
+			return err
+		}
+		obj.Object = diff.RemoveMapFields(oldObj, obj.Object)
```

Comment:
Please document the reason for all the back and forth marshaling/unmarshalling going on here.

Reply:
done